File: plugin_files/bin/scripting/timers.lua
function SetTimeout(delay, callback)
    if type(delay) ~= "number" then
        return print("The delay needs to be a number.")
    end
    if type(callback) ~= "function" then
        return print("The callback needs to be a function.")
    end

    AddTimeout(delay, callback)
end

local timerIds = 50
local timersTable = {}

function SetTimer(delay, callback)
    if type(delay) ~= "number" then
        return print("The delay needs to be a number.")
    end
    if type(callback) ~= "function" then
        return print("The callback needs to be a function.")
    end

    timerIds = timerIds + 1

    local timerId = timerIds

    timersTable[timerId] = {
        id = timerId,
        callback = callback,
        timeoutFunction = function()
            if timersTable[timerId] then
                local debuginfo = debug.getinfo(timersTable[timerId].callback, "S")
                local stackid = RegisterCallstack(GetCurrentPluginName(),
                    string.format("TimeoutCallback(path=%s,lines=%d..%d)", debuginfo.short_src, debuginfo.linedefined,
                        debuginfo.lastlinedefined))
                timersTable[timerId].callback()
                UnregisterCallstack(GetCurrentPluginName(), stackid)
                if timersTable[timerId] then
                    SetTimeout(delay, timersTable[timerId].timeoutFunction)
                end
            end
        end
    }

    SetTimeout(delay, timersTable[timerId].timeoutFunction)

    return timerId
end

function StopTimer(timerId)
    if type(timerId) ~= "number" then
        return print("The Timer ID needs to be a number.")
    end

    if not timersTable[timerId] then
        return print("Invalid Timer ID.")
    end

    timersTable[timerId] = nil
end

function NextTick(callback)
    if type(callback) ~= "function" then
        return print("The callback needs to be a function.")
    end

    SetTimeout(1, callback)
end

File: plugin_files/bin/scripting/timers.js
function LoadTimers(global) {
    let timerIds = 50
    let timersList = {}

    global.SetTimeout = (delay, callback) => {
        if(typeof delay != "number") return console.log("The delay needs to be a number.")
        if(typeof callback != "function") return console.log("The callback needs to be a function.")

        AddTimeout(delay, callback)
    }

    global.NextTick = (callback) => {
        if(typeof callback != "function") return console.log("The callback needs to be a function.")

        global.SetTimeout(1, callback)
    }

    global.StopTimer = (timerId) => {
        if (typeof timerId !== "number") {
            console.log("The Timer ID needs to be a number.");
            return;
        }
    
        if (!timersList[timerId]) {
            console.log("Invalid Timer ID.");
            return;
        }
    
        delete timersList[timerId];
    }

    global.SetTimer = (delay, callback) => {
        if(typeof delay != "number") return console.log("The delay needs to be a number.")
        if(typeof callback != "function") return console.log("The callback needs to be a function.")

        timerIds ++
        const timerId = timerIds

        timersList[timerId] = {
            id: timerId,
            callback,
            timeoutFunc: () => {
                if(timersList[timerId]) {
                    timersList[timerId].callback()
                    if(timersList[timerId]) {
                        SetTimeout(delay, timersList[timerId].timeoutFunc)
                    }
                }
            }
        }

        SetTimeout(delay, timersList[timerId].timeoutFunc)

        return timerId
    }

    global.setInterval = (callback, timeout) => {
        return SetTimer(timeout, callback)
    }
    
    global.setTimeout = (callback, timeout) => {
        return SetTimeout(timeout, callback)
    }
}

LoadTimers(globalThis)